lib/services/android-debug-service.ts

```diff
@@ -43,14 +39,60 @@ class AndroidDebugService implements IDebugService {
 			: this.debugOnDevice();
 	}
 
-	public debugOnEmulator(): IFuture<void> {
+	private debugOnEmulator(): IFuture<void> {
 		return (() => {
 			this.$platformService.deployOnEmulator(this.platform).wait();
 			this.debugOnDevice().wait();
 		}).future<void>()();
 	}
 
-	public debugOnDevice(): IFuture<void> {
+	private isPortAvailable(candidatePort: number): IFuture<boolean> {
+		let future = new Future<boolean>();
+		let server = net.createServer();
+		server.on("error", (err: Error) => {
+			future.return(false);
+		});
+		server.listen(candidatePort, (err: Error) => {
+			server.once("close", () => {
+				future.return(true);
+			});
+			server.close();
+		});
+
+		return future;
+	}
+
+	private getForwardedLocalDebugPortForPackageName(deviceId: string, packageName: string): IFuture<number> {
+		return (() => {
+			let port = -1;
+			let forwardsResult = this.device.adb.executeCommand(["forward", "--list"]).wait();
+
+			//matches 123a188909e6czzc tcp:40001 localabstract:org.nativescript.testUnixSockets-debug
+			let regexp = new RegExp(`(${deviceId} tcp:)([\\d])+(?= localabstract:${packageName}-debug)`, "g");
+			let match = forwardsResult.match(regexp);
+			if (match) {
+				port = parseInt(match[0].substring(match[0].length - 5));
+			} else {
+				let candidatePort = 40000;
+				while (!this.isPortAvailable(candidatePort++).wait()) {
+					if (candidatePort > 65534) {
+						this.$errors.failWithoutHelp("Unable to find free local port.");
+					}
+				}
+				port = candidatePort;
+
+				this.unixSocketForward(port, packageName + "-debug").wait();
+			}
+
+			return port;
+		}).future<number>()();
+	}
+
+	private unixSocketForward(local: Number, remote: String): IFuture<void> {
+		return this.device.adb.executeCommand(["forward", `tcp:${local.toString()}`, `localabstract:${remote.toString()}`]);
```

Comment:
You can safely skip the `toString()` of the `remote` variable, because it already has a string value